docs_src/design/adr/013-Device-Service-Events-Message-Bus.md

```diff
@@ -0,0 +1,215 @@
+
+The App Service SDK will be enhanced for the secure Message Bus connection described below. See **[Secure Connections](#secure-connections)** for details
+
+### Message Bus Topics
+
+> *Note: The change recommended here is not required for this design, but it provides a good opportunity to adopt it.*
+
+Currently Core Data publishes Events to the simple `events` topic. All Application Services running receive all Events published, whether they want them or not. The Events can be filtered out using the `FilterByDeviceName` pipeline function, but the Application Services still receive all the Events and process all the Events to some extent. This could cause load issues in a deployment with many devices and large volume of Events.
+
+Pub/Sub systems have advanced topic schema, which we can take advantage of to filter for just the Events the Application Service actual needs. If the publishers of Events add the `Device Name` to the topic in the form `edgex/events/<device-name>` then the Application Service can filter for just the Events from the device(s) it wants by only subscribing to those `Device Names`, i.e. `edgex/events/Random-Integer-Device` . If persistence is require, Core Data will subscribe using the `#` wild card, i.e. `edgex/events/#` , so that it receives all Events. 
+
+The Message Bus abstraction allows for multiple subscriptions, so an Application Service could specify to receive data from only specific devices by creating multiple subscriptions. i.e.  `edgex/Events/Random-Integer-Device` and  `edgex/Events/Random-Boolean-Device`. Currently the App SDK only allows for a single subscription topic to be configured, but that could easily be expanded to handle a list of subscriptions. See [Configuration](#configuration) section below for details. 
+
+Core Data's existing publishing of Events would also need to be changed to use this new topic schema. One challenge with this is Core Data doesn't currently know the `Device Name` when it receives a CBOR encoded event. This is because it doesn't decode the Event until after it publishes it to the Message Bus. The V2 API could be enhanced to require the `Device Name` in the HTTP header when content type is CBOR.
+
+### Configuration
+
+#### Device Services
+
+All Device services will have the following additional configuration to allow connecting and publishing to the Message Bus.
+
+##### [MessageQueue]
+
+A  MessageQueue section will be added, which is similar to that used in Core Data today. To enable secure connections, the `Username` & `Password` have been replaced with `Authmode` & `SecretPath`, See **[Secure Connections](#secure-connections)** section below for details. The added `Enabled` property controls whether the Device Service publishes to the Message Bus or POSTs to Core Data. 
+
+```toml
+[MessageQueue]
+Enabled = true
+Protocol = 'tcp'
+Host = 'localhost'
+Port = 1883
+Type = 'mqtt'
+PublishTopic = 'edgex/events/<device-name>'
```

Comment:
Presume this would have to be a list if a service was responsible for more than one device? Or is this the template and the app fills in the "device-name".  Of course this presumes that the device name is unique across the EdgeX instance (which I believe is still the rule).

Reply:
As stated in other response, I now think this should be `edgex/events/<device-profile>` so we are filtering on a class of devices, not on an instance. We could then take it one-step future to allow filter on a specific instance from a device profile by using this schema `edgex/events/<device-profile>/<device-name>` which leaves it open to these scenarios for publishing to theses topics:

1. - `edgex/events` - No place holder for <device-profile> or <device-name> specified
2. - `edgex/events/Random-Integer-Device` - Only place holder for <device-profile> specified
3. - `edgex/events/Random-Integer-Device/Random-Integer-Device` - Both place holders specified.

Subscribers could have any of these subscribe topics:
-  `edgex/events` - only receive Events for 1. above 
-  `edgex/events/#` - Receive Events for all above 
-  `edgex/events/Random-Integer-Device/#` - Receive Events for 2. & 3. above 
-  `edgex/events/Random-Integer-Device/Random-Integer-Device` - Receive Events for only  3. above

Reply:
Suggest this should be non-configurable at the DS, ie the publishing topic is always edgex/events/dev-profile/dev-name.
We could reflect this in the v2 REST endpoint for publishing to core-data? ie POST to api/v2/events/<profile>/<name> and this would remove the need for sending additional information outside of a CBOR payload

Reply:
Not sure about not allowing it to be configured, but do like the idea of adding `Profile` and `Name` to the rest path for V2 Add Event API. Seems cleaner than putting it in the HTTP Header.